File: Problems/Algorithms/130. Surrounded Regions/surrounded_regions.go
func solve(board [][]byte)  {
    n, m := len(board), len(board[0])

    neighbors := [4][2]int{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}

    queue := make([][2]int, 0)
    for i := 0; i < n; i++ {
        if board[i][0] == 'O' {
            queue = append(queue, [2]int{i, 0})
        }
        if board[i][m-1] == 'O' {
            queue = append(queue, [2]int{i, m-1})
        }
    }
    for j := 0; j < m; j++ {
        if board[0][j] == 'O' {
            queue = append(queue, [2]int{0, j})
        }
        if board[n-1][j] == 'O' {
            queue = append(queue, [2]int{n-1, j})
        }
    }

    for len(queue) > 0 {
        curr := queue[0]
        queue = queue[1:]
        r, c := curr[0], curr[1]
        board[r][c] = 'B'

        for _, neighbor := range neighbors {
            x, y := r + neighbor[0], c + neighbor[1]
            if !isValid(n, m, x, y) {
                continue
            }

            if (board[x][y] == 'X' || board[x][y] == 'B') {
                continue
            }

            queue = append(queue, [2]int{x, y})
        }
    }

    for i := 0; i < n; i++ {
        for j := 0; j < m; j++ {
            if board[i][j] == 'O' {
                board[i][j] = 'X'
            } else if board[i][j] == 'B' {
                board[i][j] = 'O'
            }
        }
    }
}

func isValid(n int, m int, r int, c int) bool {
    return r >= 0 && r < n && c >= 0 && c < m
}

File: Problems/Algorithms/130. Surrounded Regions/surrounded_regions.rs
use std::collections::VecDeque;

impl Solution {
    pub fn solve(board: &mut Vec<Vec<char>>) {
        let (n, m) = (board.len(), board[0].len());
        let neighbors = [[0, 1], [0, -1], [1, 0], [-1, 0]];
        
        let mut queue: VecDeque<(usize, usize)> = VecDeque::new();
        for i in 0..n {
            if board[i][0] == 'O' {
                queue.push_back((i, 0));
            }
            if board[i][m-1] == 'O' {
                queue.push_back((i, m-1));
            }
        }
        for j in 0..m {
            if board[0][j] == 'O' {
                queue.push_back((0, j));
            }
            if board[n-1][j] == 'O' {
                queue.push_back((n-1, j));
            }
        }

        while let Some(curr) = queue.pop_front() {
            let (r, c) = curr;
            board[r][c] = 'B';
            
            for neighbor in neighbors.iter() {
                let (x, y) = (r + neighbor[0] as usize, c + neighbor[1] as usize);
                if x >= 0 && x < n && y >= 0 && y < m && board[x][y] == 'O' {
                    queue.push_back((x, y));
                }
            }
        }
        
        for row in board.iter_mut() {
            for col in row.iter_mut() {
                *col = match *col {
                    'O' => 'X',
                    'B' => 'O',
                    col => col,
                }
            }
        }
    }
}

